dattri/algorithm/trak.py

```diff
@@ -108,6 +125,21 @@ def cache(
                 grad_t = self.grad_func(parameters, train_batch_data)
                 grad_t = torch.nan_to_num(grad_t)
                 grad_t /= self.norm_scaler
+
+                # Unflatten the gradient tensors
+                grad_t = [
+                    _unflatten_params(g, self.task.model)
+                    for g in grad_t
+                ]
+
+                # Stack the gradient tensors for each layer
+                grad_t = {
+                    key: torch.stack(
+                        [g[key] for g in grad_t],
+                        ).view(len(grad_t), -1)
+                    for key in grad_t[0]
+                }
```

Comment:
I don't think we need this step? Eventually all the params (in each module) will be projected and added together.

Reply:
I think it's just a convenient way to project batched gradients. Here we just stacked the gradients in each layer as `(batch_size, dim_out * dim_in)` so that the projector can process them all at once. Otherwise we may need to use a loop to process every gradient in the batch one by one.